### ISSUE TITLE
schema calls `deep_symbolize` turning everything into hashes

### ISSUE THREAD

**Issue by kaikuchn** [hanami/model#394]
Hey folks,

I have this entity
```ruby
class AlertRule < Hanami::Entity
  attributes do
    attribute :id,        Types::Coercible::Int
    attribute :topic,     Types::Coercible::String
    attribute :criterion, Types::AlertCriterion
    attribute :action,    Types::AlertAction
  end
end
```

where `Types::AlertCriterion` and `Types::AlertAction` are custom types which coerce the input to instances of `AlertCriterion` and `AlertAction` respectively. Those are Virtus value objects, since I haven't yet switched them to dry-structs or hanami-entities yet, and respond to `to_hash`. Therefore when I instantiate an `AlertRule` the schema hashifies criterion and action, due to [this line of code](https://github.com/hanami/model/blob/f8324cb881822d03e18de6d08cec1bdabf74e972/lib/hanami/entity/schema.rb#L226).

Is there a specific reason for the deep symbolizing? I would argue that shallow symbolizing should suffice since any hash type would have symbolized during it's own schema application already and any non-hash should probably stay a non-hash. Otherwise I'd have specified a different type..?
What do you think?

**Comment by jodosha** [comment]
@kaikuchn That is a problem that also @mereghost stumbled on while working on associations. Can you folks please explore if we can replace `deep_symbolize` with `symbolize`?

**Comment by kaikuchn** [comment]
Have a look at the proposed change: https://github.com/hanami/model/pull/395
This would keep deep symbolising for schema-less entities but leave it up to the user to specify whether or not he wishes to symbolise nested hashes through use of `Type::Hash.symbolized`.
Automatic schema uses `schema` instead of `symbolized` though. Since `symbolized` is based on `weak` we cannot simply substitute this.

**Comment by kaikuchn** [comment]
After investigating how automatic schema inference works I've adapted the linked PR to deeply symbolise hashes for inferred/automatic schemas.

**Comment by jodosha** [comment]
Fixed by #395 and #401
